File: apps/dsa-web/src/types/analysis.ts
/**
 * Analysis-related type definitions.
 * Aligned with the API schema.
 */

// ============ Request Types ============

export interface AnalysisRequest {
  stockCode?: string;
  stockCodes?: string[];
  reportType?: 'simple' | 'detailed' | 'full' | 'brief';
  forceRefresh?: boolean;
  asyncMode?: boolean;
  stockName?: string;
  originalQuery?: string;
  selectionSource?: 'manual' | 'autocomplete' | 'import' | 'image';
}

// ============ Report Types ============

export type ReportLanguage = 'zh' | 'en';

/** Report metadata */
export interface ReportMeta {
  id?: number;  // Analysis history record ID, present for persisted reports
  queryId: string;
  stockCode: string;
  stockName: string;
  reportType: 'simple' | 'detailed' | 'full' | 'brief';
  reportLanguage?: ReportLanguage;
  createdAt: string;
  marketTimestamp?: string;
  marketSessionDate?: string;
  newsPublishedAt?: string;
  reportGeneratedAt?: string;
  currentPrice?: number;
  changePct?: number;
  modelUsed?: string;  // LLM model used for analysis
}

/** Sentiment label */
export type SentimentLabel =
  | '极度悲观'
  | '悲观'
  | '中性'
  | '乐观'
  | '极度乐观'
  | 'Very Bearish'
  | 'Bearish'
  | 'Neutral'
  | 'Bullish'
  | 'Very Bullish';

/** Report summary section */
export interface ReportSummary {
  analysisSummary: string;
  operationAdvice: string;
  trendPrediction: string;
  sentimentScore: number;
  sentimentLabel?: SentimentLabel;
}

/** Strategy section */
export interface ReportStrategy {
  idealBuy?: string;
  secondaryBuy?: string;
  stopLoss?: string;
  takeProfit?: string;
}

export interface StandardReportField {
  label: string;
  value: string;
  source?: string;
  status?: string;
  category?: string;
}

export interface StandardReportTag {
  label: string;
  value: string;
}

export interface StandardReportSummaryPanel {
  stock?: string;
  ticker?: string;
  score?: number;
  currentPrice?: string;
  changeAmount?: string;
  changePct?: string;
  marketTime?: string;
  marketSessionDate?: string;
  sessionLabel?: string;
  operationAdvice?: string;
  trendPrediction?: string;
  oneSentence?: string;
  timeSensitivity?: string;
  tags?: StandardReportTag[];
}

export interface StandardReportRegularMetrics {
  price?: number;
  prevClose?: number;
  changeAmount?: number;
  changePct?: number;
  open?: number;
  high?: number;
  low?: number;
  close?: number;
  amplitude?: number;
  volume?: number;
  amount?: number;
  volumeRatio?: number;
  turnoverRate?: number;
  averagePrice?: number;
  vwap?: number;
}

export interface StandardReportMarketBlock {
  regularFields?: StandardReportField[];
  extendedFields?: StandardReportField[];
  consistencyWarnings?: string[];
  regularPriceNumeric?: number;
  regularMetrics?: StandardReportRegularMetrics;
  timeContext?: Record<string, string | undefined>;
}

export interface StandardReportTableSection {
  title: string;
  fields: StandardReportField[];
  note?: string;
}

export interface StandardReportVisualScore {
  value?: number;
  max?: number;
}

export interface StandardReportPricePosition {
  currentPrice?: number;
  ma20?: number;
  ma60?: number;
  distanceToMa20Pct?: number;
  distanceToMa60Pct?: number;
  vsMa20?: string;
  vsMa60?: string;
}

export interface StandardReportRiskOpportunity {
  positiveCount?: number;
  riskCount?: number;
  latestNewsCount?: number;
}

export interface StandardReportTrendStrength {
  value?: number;
  max?: number;
  label?: string;
}

export interface StandardReportVisualBlocks {
  score?: StandardReportVisualScore;
  trendStrength?: StandardReportTrendStrength;
  pricePosition?: StandardReportPricePosition;
  riskOpportunity?: StandardReportRiskOpportunity;
}

export interface StandardReportHighlights {
  positiveCatalysts?: string[];
  riskAlerts?: string[];
  latestNews?: string[];
  newsValueGrade?: string;
  sentimentSummary?: string;
  earningsOutlook?: string;
}

export interface StandardReportBattlePlanItem {
  label: string;
  value: string;
  tone?: string;
}

export interface StandardReportBattlePlanCompact {
  cards?: StandardReportBattlePlanItem[];
  notes?: StandardReportBattlePlanItem[];
  warnings?: string[];
}

export interface StandardReportDecisionContext {
  shortTermView?: string;
  compositeView?: string;
  adjustmentReason?: string;
  changeReason?: string;
  previousScore?: string;
  scoreChange?: string;
  scoreBreakdown?: StandardReportScoreBreakdownItem[];
}

export interface StandardReportScoreBreakdownItem {
  label: string;
  score?: number;
  note?: string;
  tone?: 'default' | 'success' | 'warning' | 'danger' | 'info' | 'history' | string;
}

export interface StandardReportChecklistItem {
  status: 'pass' | 'warn' | 'fail' | 'info' | 'na' | string;
  icon: string;
  text: string;
}

export interface StandardReport {
  title?: Record<string, unknown>;
  summaryPanel?: StandardReportSummaryPanel;
  infoFields?: StandardReportField[];
  positionAdvice?: Record<string, string | undefined>;
  market?: StandardReportMarketBlock;
  technicalFields?: StandardReportField[];
  fundamentalFields?: StandardReportField[];
  earningsFields?: StandardReportField[];
  sentimentFields?: StandardReportField[];
  tableSections?: {
    market?: StandardReportTableSection;
    technical?: StandardReportTableSection;
    fundamental?: StandardReportTableSection;
    earnings?: StandardReportTableSection;
  };
  visualBlocks?: StandardReportVisualBlocks;
  decisionContext?: StandardReportDecisionContext;
  highlights?: StandardReportHighlights;
  battleFields?: StandardReportField[];
  battlePlanCompact?: StandardReportBattlePlanCompact;
  battleWarnings?: string[];
  checklist?: string[];
  checklistItems?: StandardReportChecklistItem[];
}

/** Details section */
export interface ReportDetails {
  newsContent?: string;
  rawResult?: Record<string, unknown>;
  contextSnapshot?: Record<string, unknown>;
  standardReport?: StandardReport;
  financialReport?: Record<string, unknown>;
  dividendMetrics?: Record<string, unknown>;
}

/** Full analysis report */
export interface AnalysisReport {
  meta: ReportMeta;
  summary: ReportSummary;
  strategy?: ReportStrategy;
  details?: ReportDetails;
}

// ============ Analysis Result Types ============

/** Sync analysis response */
export interface AnalysisResult {
  queryId: string;
  stockCode: string;
  stockName: string;
  report: AnalysisReport;
  createdAt: string;
}

/** Async task accepted response */
export interface TaskAccepted {
  taskId: string;
  status: 'pending' | 'processing';
  message?: string;
}

export interface BatchTaskAcceptedItem {
  taskId: string;
  stockCode: string;
  status: 'pending' | 'processing';
  message?: string;
}

export interface BatchDuplicateTaskItem {
  stockCode: string;
  existingTaskId: string;
  message: string;
}

export interface BatchTaskAcceptedResponse {
  accepted: BatchTaskAcceptedItem[];
  duplicates: BatchDuplicateTaskItem[];
  message: string;
}

export type AnalyzeAsyncResponse = TaskAccepted | BatchTaskAcceptedResponse;

export type AnalyzeResponse = AnalysisResult | AnalyzeAsyncResponse;

/** Task status */
export interface TaskStatus {
  taskId: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  progress?: number;
  result?: AnalysisResult;
  error?: string;
  stockName?: string;
  originalQuery?: string;
  selectionSource?: string;
}

/** Task details used by task list and SSE events */
export interface TaskInfo {
  taskId: string;
  stockCode: string;
  stockName?: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  progress: number;
  message?: string;
  reportType: string;
  createdAt: string;
  startedAt?: string;
  completedAt?: string;
  error?: string;
  originalQuery?: string;
  selectionSource?: string;
}

/** Task list response */
export interface TaskListResponse {
  total: number;
  pending: number;
  processing: number;
  tasks: TaskInfo[];
}

/** Duplicate task error response */
export interface DuplicateTaskError {
  error: 'duplicate_task';
  message: string;
  stockCode: string;
  existingTaskId: string;
}

// ============ History Types ============

/** History item summary */
export interface HistoryItem {
  id: number;  // Record primary key ID, always present for persisted history items
  queryId: string;  // Linked analysis query ID
  stockCode: string;
  stockName?: string;
  reportType?: string;
  sentimentScore?: number;
  operationAdvice?: string;
  createdAt: string;
}

/** History list response */
export interface HistoryListResponse {
  total: number;
  page: number;
  limit: number;
  items: HistoryItem[];
}

/** News item */
export interface NewsIntelItem {
  title: string;
  snippet: string;
  url: string;
}

/** News response */
export interface NewsIntelResponse {
  total: number;
  items: NewsIntelItem[];
}

/** History filter parameters */
export interface HistoryFilters {
  stockCode?: string;
  startDate?: string;
  endDate?: string;
}

/** History pagination parameters */
export interface HistoryPagination {
  page: number;
  limit: number;
}

// ============ Error Types ============

export interface ApiError {
  error: string;
  message: string;
  detail?: Record<string, unknown>;
}

// ============ Helper Functions ============

/** Get sentiment label by score */
export const getSentimentLabel = (score: number, language: ReportLanguage = 'zh'): SentimentLabel => {
  if (language === 'en') {
    if (score <= 20) return 'Very Bearish';
    if (score <= 40) return 'Bearish';
    if (score <= 60) return 'Neutral';
    if (score <= 80) return 'Bullish';
    return 'Very Bullish';
  }
  if (score <= 20) return '极度悲观';
  if (score <= 40) return '悲观';
  if (score <= 60) return '中性';
  if (score <= 80) return '乐观';
  return '极度乐观';
};

/** Get sentiment color by score */
export const getSentimentColor = (score: number): string => {
  if (score <= 20) return '#ef4444'; // red-500
  if (score <= 40) return '#f97316'; // orange-500
  if (score <= 60) return '#eab308'; // yellow-500
  if (score <= 80) return '#22c55e'; // green-500
  return '#10b981'; // emerald-500
};

File: api/v1/schemas/analysis.py
# -*- coding: utf-8 -*-
"""Analysis API schemas."""

from enum import Enum
from typing import Any, List, Optional

from pydantic import AliasChoices, BaseModel, ConfigDict, Field

from src.utils.analysis_metadata import SELECTION_SOURCE_PATTERN


class TaskStatusEnum(str, Enum):
    """Task status values."""

    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


class AnalyzeRequest(BaseModel):
    """Analysis request parameters."""

    stock_code: Optional[str] = Field(None, description="단일 종목 코드", json_schema_extra={"example": "600519"})
    stock_codes: Optional[List[str]] = Field(None, description="여러 종목 코드. stock_code와 둘 중 하나만 사용합니다.", json_schema_extra={"example": ["600519", "000858"]})
    report_type: str = Field("detailed", description="보고서 유형: simple(간단) / detailed(상세) / full(전체) / brief(요약)", pattern="^(simple|detailed|full|brief)$")
    force_refresh: bool = Field(False, description="캐시를 무시하고 강제로 새로 분석할지 여부")
    async_mode: bool = Field(False, description="비동기 작업으로 실행할지 여부")
    stock_name: Optional[str] = Field(None, description="사용자가 선택한 종목명", json_schema_extra={"example": "贵州茅台"})
    original_query: Optional[str] = Field(None, description="사용자의 원본 입력", json_schema_extra={"example": "茅台"})
    selection_source: Optional[str] = Field(None, description="종목 선택 출처", pattern=SELECTION_SOURCE_PATTERN, json_schema_extra={"example": "autocomplete"})
    notify: bool = Field(True, description="푸시 알림을 보낼지 여부")
    skills: Optional[List[str]] = Field(
        None,
        validation_alias=AliasChoices("skills", "strategies"),
        description="이번 분석에 사용할 strategy skill ID 목록. legacy strategies 필드와 호환됩니다.",
        json_schema_extra={"example": ["bull_trend", "growth_quality"]},
    )

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "stock_code": "600519",
            "report_type": "detailed",
            "force_refresh": False,
            "async_mode": False,
            "stock_name": "贵州茅台",
            "original_query": "茅台",
            "selection_source": "autocomplete",
            "notify": True,
            "skills": ["bull_trend"],
        }
    })


class MarketReviewRequest(BaseModel):
    """Market review trigger parameters."""

    send_notification: bool = Field(True, description="시장 리뷰 완료 후 푸시 알림을 보낼지 여부")


class MarketReviewAccepted(BaseModel):
    """Market review background task accepted response."""

    status: str = Field("accepted", description="제출 상태")
    message: str = Field(..., description="안내 메시지")
    send_notification: bool = Field(..., description="알림 발송 여부")
    task_id: Optional[str] = Field(None, description="작업이 실제로 제출된 경우 반환되는 작업 ID")


class AnalysisResultResponse(BaseModel):
    """Analysis result response."""

    query_id: str = Field(..., description="분석 기록 고유 ID")
    stock_code: str = Field(..., description="종목 코드")
    stock_name: Optional[str] = Field(None, description="종목명")
    report: Optional[Any] = Field(None, description="분석 보고서")
    created_at: str = Field(..., description="생성 시간")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "query_id": "abc123def456",
            "stock_code": "AAPL",
            "stock_name": "Apple",
            "report": {"summary": {"sentiment_score": 75, "operation_advice": "hold"}},
            "created_at": "2024-01-01T12:00:00",
        }
    })


class TaskAccepted(BaseModel):
    """Async task accepted response."""

    task_id: str = Field(..., description="상태 조회에 사용할 작업 ID")
    status: str = Field(..., description="작업 상태", pattern="^(pending|processing)$")
    message: Optional[str] = Field(None, description="안내 메시지")

    model_config = ConfigDict(json_schema_extra={
        "example": {"task_id": "task_abc123", "status": "pending", "message": "Analysis task accepted"}
    })


class BatchTaskAcceptedItem(BaseModel):
    """Successfully submitted item in a batch async task request."""

    task_id: str = Field(..., description="상태 조회에 사용할 작업 ID")
    stock_code: str = Field(..., description="종목 코드")
    status: str = Field(..., description="작업 상태", pattern="^(pending|processing)$")
    message: Optional[str] = Field(None, description="안내 메시지")

    model_config = ConfigDict(json_schema_extra={
        "example": {"task_id": "task_abc123", "stock_code": "AAPL", "status": "pending", "message": "분석 작업이 큐에 추가되었습니다: AAPL"}
    })


class BatchDuplicateTaskItem(BaseModel):
    """Duplicate item in a batch async task request."""

    stock_code: str = Field(..., description="종목 코드")
    existing_task_id: str = Field(..., description="이미 존재하는 작업 ID")
    message: str = Field(..., description="오류 메시지")

    model_config = ConfigDict(json_schema_extra={
        "example": {"stock_code": "AAPL", "existing_task_id": "task_existing_123", "message": "종목 AAPL이 이미 분석 중입니다. (task_id: task_existing_123)"}
    })


class BatchTaskAcceptedResponse(BaseModel):
    """Batch async task accepted response."""

    accepted: List[BatchTaskAcceptedItem] = Field(default_factory=list, description="성공적으로 제출된 작업 목록")
    duplicates: List[BatchDuplicateTaskItem] = Field(default_factory=list, description="중복으로 건너뛴 작업 목록")
    message: str = Field(..., description="요약 메시지")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "accepted": [{"task_id": "task_abc123", "stock_code": "AAPL", "status": "pending", "message": "분석 작업이 큐에 추가되었습니다: AAPL"}],
            "duplicates": [{"stock_code": "MSFT", "existing_task_id": "task_existing_456", "message": "종목 MSFT가 이미 분석 중입니다. (task_id: task_existing_456)"}],
            "message": "1개 작업을 제출했고 1개 중복 작업은 건너뛰었습니다.",
        }
    })


class TaskStatus(BaseModel):
    """Task status model."""

    task_id: str = Field(..., description="작업 ID")
    status: str = Field(..., description="작업 상태", pattern="^(pending|processing|completed|failed)$")
    progress: Optional[int] = Field(None, description="진행률(0-100)", ge=0, le=100)
    result: Optional[AnalysisResultResponse] = Field(None, description="분석 결과. completed 상태에서만 존재합니다.")
    market_review_report: Optional[str] = Field(None, description="시장 리뷰 작업이 반환한 보고서 텍스트")
    error: Optional[str] = Field(None, description="오류 메시지. failed 상태에서만 존재합니다.")
    stock_name: Optional[str] = Field(None, description="종목명")
    original_query: Optional[str] = Field(None, description="사용자의 원본 입력")
    selection_source: Optional[str] = Field(None, description="선택 출처", pattern=SELECTION_SOURCE_PATTERN)
    skills: Optional[List[str]] = Field(None, description="이번 작업에 사용한 strategy skill ID 목록")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "task_id": "task_abc123",
            "status": "completed",
            "progress": 100,
            "result": None,
            "market_review_report": None,
            "error": None,
            "stock_name": "Apple",
            "original_query": "Apple",
            "selection_source": "autocomplete",
            "skills": ["bull_trend"],
        }
    })


class TaskInfo(BaseModel):
    """Task details model used for task lists and SSE events."""

    task_id: str = Field(..., description="작업 ID")
    stock_code: str = Field(..., description="종목 코드")
    stock_name: Optional[str] = Field(None, description="종목명")
    status: TaskStatusEnum = Field(..., description="작업 상태")
    progress: int = Field(0, description="진행률(0-100)", ge=0, le=100)
    message: Optional[str] = Field(None, description="상태 메시지")
    report_type: str = Field("detailed", description="보고서 유형")
    created_at: str = Field(..., description="생성 시간")
    started_at: Optional[str] = Field(None, description="시작 시간")
    completed_at: Optional[str] = Field(None, description="완료 시간")
    error: Optional[str] = Field(None, description="오류 메시지. failed 상태에서만 존재합니다.")
    original_query: Optional[str] = Field(None, description="사용자의 원본 입력")
    selection_source: Optional[str] = Field(None, description="선택 출처", pattern=SELECTION_SOURCE_PATTERN)
    skills: Optional[List[str]] = Field(None, description="이번 작업에 사용한 strategy skill ID 목록")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "task_id": "abc123def456",
            "stock_code": "AAPL",
            "stock_name": "Apple",
            "status": "processing",
            "progress": 50,
            "message": "분석 중...",
            "report_type": "detailed",
            "created_at": "2026-02-05T10:30:00",
            "started_at": "2026-02-05T10:30:01",
            "completed_at": None,
            "error": None,
            "original_query": "Apple",
            "selection_source": "autocomplete",
            "skills": ["bull_trend"],
        }
    })


class TaskListResponse(BaseModel):
    """Task list response."""

    total: int = Field(..., description="전체 작업 수")
    pending: int = Field(..., description="대기 중인 작업 수")
    processing: int = Field(..., description="처리 중인 작업 수")
    tasks: List[TaskInfo] = Field(..., description="작업 목록")

    model_config = ConfigDict(json_schema_extra={
        "example": {"total": 3, "pending": 1, "processing": 2, "tasks": []}
    })


class DuplicateTaskErrorResponse(BaseModel):
    """Duplicate task error response."""

    error: str = Field("duplicate_task", description="오류 유형")
    message: str = Field(..., description="오류 메시지")
    stock_code: str = Field(..., description="종목 코드")
    existing_task_id: str = Field(..., description="이미 존재하는 작업 ID")

    model_config = ConfigDict(json_schema_extra={
        "example": {
            "error": "duplicate_task",
            "message": "종목 AAPL이 이미 분석 중입니다.",
            "stock_code": "AAPL",
            "existing_task_id": "abc123def456",
        }
    })
